unittest/test_diff_actions.cpp

```diff
@@ -20,29 +20,29 @@ using namespace crocoddyl::unittest;
 void test_check_data(DifferentialActionModelTypes::Type action_type) {
   // create the model
   DifferentialActionModelFactory factory;
-  const boost::shared_ptr<crocoddyl::DifferentialActionModelAbstract>& model = factory.create(action_type);
+  boost::shared_ptr<crocoddyl::DifferentialActionModelAbstract> model = factory.create(action_type);
```

Comment:
Is this triggered issues in clang?

I wonder why do we need to this?

Reply:
I changed it during debugging. Note that the factory returns the shared_ptr and thus whoever calls `create` should own it. But we would only store a const-ref to the pointer. What we could do is return a unique_ptr from the factory thus signifying that whoever calls it needs to take ownership (and we can instantiate a shared from a unique_ptr)

Reply:
It is not a good idea to use  an `unique_ptr` as we (something) need to use these pointers for constructing other Crocoddyl objects.

The proposed code is creating a second owner.
However, the current approach avoids it by passing a reference.
Note that multiple owners add extra (but tiny) runtime overhead.

My question is **does boost unit-test need this?**

Reply:
We can have methods that create objects return unique_ptr - if we need shared_ptr, the one instantiating can create a shared_ptr from a unique_ptr:
```cpp
std::unique_ptr<Something> createSomething() { return std::make_unique<Something>(); }

std::shared_ptr<Something> something = createSomething();
```

Cf. https://godbolt.org/z/xvf41ah6e

> The proposed code is creating a second owner.

I am not sure it is given return-value optimisation. The `create` method in this case does not keep a shared_ptr itself.

The change was part of the debugging to resolve the Clang unittest crashes.

@jcarpent Any thoughts?